Error in step 4 during upgrade to 2.1.0rc4

After doing the 4th step (`make -f v2.mk upgrade4`), I get the following error : 
```
 - /layers/plan_ville_vevey_1986/dimensions/0: Cannot find required key 'generate'.
 - /layers/plan_ville_vevey_1986/dimensions/0: Key 'value' was not defined.
 - /layers/plan_ville_vevey_1993/dimensions/0: Cannot find required key 'generate'.
 - /layers/plan_ville_vevey_1993/dimensions/0: Key 'value' was not defined.
 - /layers/plan_ville_vevey_2000/dimensions/0: Cannot find required key 'generate'.
 - /layers/plan_ville_vevey_2000/dimensions/0: Key 'value' was not defined.
 - /layers/plan_ville_voirie/dimensions/0: Cannot find required key 'generate'.
 - /layers/plan_ville_voirie/dimensions/0: Key 'value' was not defined.
 - /layers/rayonnement_solaire_2012/dimensions/0: Cannot find required key 'generate'.
 - /layers/rayonnement_solaire_2012/dimensions/0: Key 'value' was not defined.
 - /layers/vevey_moyen_age/dimensions/0: Cannot find required key 'generate'.
 - /layers/vevey_moyen_age/dimensions/0: Key 'value' was not defined.
CONST_Makefile:1148: recipe for target 'apache/mapcache.xml' failed
make[1]: *** [apache/mapcache.xml] Error 1
make[1]: Leaving directory '/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish'
Traceback (most recent call last):
  File ".build/venv/bin/c2ctool", line 11, in <module>
    sys.exit(main())
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 98, in main
    c2ctool.upgrade()
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 222, in upgrade
    self.step4()
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 442, in step4
    check_call(["make", "-f", self.options.file, "build"])
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make', '-f', 'v2.mk', 'build']' returned non-zero exit status 2
CONST_Makefile:1256: recipe for target 'upgrade4' failed
make: *** [upgrade4] Error 1
```

Ok thanks, now I have that 

        dimensions:
            -   name: DATE
                # the default value for the WMTS capabilities
                default: "2011"
                # the generated value
                generate: "2011"
                # all the available values in the WMTS capabilities
                values: ["2011"]

and the following error : 

CONST_Makefile:1148: warning: overriding recipe for target 'apache/mapcache.xml'
cartoriviera.mk:18: warning: ignoring old recipe for target 'apache/mapcache.xml'
.build/venv/bin/c2ctool upgrade  v2.mk --step 4
make[1]: Entering directory '/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish'
CONST_Makefile:1148: warning: overriding recipe for target 'apache/mapcache.xml'
cartoriviera.mk:18: warning: ignoring old recipe for target 'apache/mapcache.xml'
.build/venv/bin/c2c-template --vars vars_prod_v2.yaml --engine mako --files tilegeneration/config.yaml.mako
.build/venv/bin/generate_controller --generate-mapcache-config
ERROR:tilecloud_chain:__init__:The config file 'tilegeneration/config.yaml' in invalid.
 - /caches/local/http_urls/0: Value 'https://preprod.cartoriviera.ch/tiles' does not match pattern '^https?://.+/$'.
CONST_Makefile:1148: recipe for target 'apache/mapcache.xml' failed
make[1]: *** [apache/mapcache.xml] Error 1
make[1]: Leaving directory '/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish'
Traceback (most recent call last):
  File ".build/venv/bin/c2ctool", line 11, in <module>
    sys.exit(main())
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 98, in main
    c2ctool.upgrade()
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 222, in upgrade
    self.step4()
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 442, in step4
    check_call(["make", "-f", self.options.file, "build"])
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make', '-f', 'v2.mk', 'build']' returned non-zero exit status 2
CONST_Makefile:1256: recipe for target 'upgrade4' failed
make: *** [upgrade4] Error 1

Ok I have added a / in my config.yaml.mako (strange, it was not needed in previous rc...)

caches:
    local:
        type: filesystem
        folder: /var/sig/tilecache/cartoriviera
        wmtscapabilities_file: 1.0.0/WMTSCapabilities_${instanceid}.xml
        # for GetCapabilities
        # can includes %(folder)s comes from this cache
        http_urls: 
        - https://${host}/tiles/

Now I have the folowing error.... : 

.build/venv/bin/generate_controller --generate-apache-config
.build/venv/bin/c2c-template --vars vars_prod_v2.yaml --engine mako --files print/templates/A0_landscape.mako
.build/venv/bin/c2c-template --vars vars_prod_v2.yaml --engine mako --files print/templates/A1_landscape.mako
.build/venv/bin/c2c-template --vars vars_prod_v2.yaml --engine mako --files print/templates/A4_portrait.mako
.build/venv/bin/c2c-template --vars vars_prod_v2.yaml --engine mako --files print/templates/A4_portrait_inherit.mako
Traceback (most recent call last):
  File ".build/venv/bin/c2c-template", line 11, in <module>
    sys.exit(main())
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2c/template.py", line 213, in main
    _proceed(files, used_vars, options)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2c/template.py", line 237, in _proceed
    bottle_template(files, used_vars, engine)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2c/template.py", line 268, in bottle_template
    template, **used_vars
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/bin/bottle.py", line 3609, in template
    return TEMPLATES[tplid].render(kwargs)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/bin/bottle.py", line 3280, in render
    return self.tpl.render(**_defaults)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/mako/template.py", line 462, in render
    return runtime._render(self, self.callable_, args, data)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/mako/runtime.py", line 838, in _render
    **_kwargs_for_callable(callable_, data))
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/mako/runtime.py", line 872, in _render_context
    (inherit, lclcontext) = _populate_self_namespace(context, tmpl)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/mako/runtime.py", line 814, in _populate_self_namespace
    ret = template.module._mako_inherit(template, context)
  File "print_templates_A4_portrait_inherit_mako", line 25, in _mako_inherit
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/mako/runtime.py", line 771, in _inherit_from
    template = _lookup_template(context, uri, calling_uri)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/mako/runtime.py", line 804, in _lookup_template
    raise exceptions.TemplateLookupException(str(compat.exception_as()))
mako.exceptions.TemplateLookupException: Cant locate template for uri u'/A4_portrait.mako'
CONST_Makefile:650: recipe for target 'print/templates/A4_portrait_inherit' failed
make[1]: *** [print/templates/A4_portrait_inherit] Error 1
make[1]: Leaving directory '/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish'
Traceback (most recent call last):
  File ".build/venv/bin/c2ctool", line 11, in <module>
    sys.exit(main())
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 98, in main
    c2ctool.upgrade()
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 222, in upgrade
    self.step4()
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 442, in step4
    check_call(["make", "-f", self.options.file, "build"])
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['make', '-f', 'v2.mk', 'build']' returned non-zero exit status 2
CONST_Makefile:1256: recipe for target 'upgrade4' failed
make: *** [upgrade4] Error 1

I tried without the first slash and it is OK for that error.  Now I have the following error. It tries to set to not null a column that contains null values... Do I have to run BD migration manually before ? 

INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade ee25d267bf46 -> e004f76e951a, Add missing not null
Traceback (most recent call last):
  File ".build/venv/bin/c2ctool", line 11, in <module>
    sys.exit(main())
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 98, in main
    c2ctool.upgrade()
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 222, in upgrade
    self.step4()
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/c2cgeoportal/scripts/c2ctool.py", line 444, in step4
    command.upgrade(Config("alembic.ini"), "head")
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/command.py", line 174, in upgrade
    script.run_env()
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/script/base.py", line 407, in run_env
    util.load_python_file(self.dir, 'env.py')
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/util/pyfiles.py", line 93, in load_python_file
    module = load_module_py(module_id, path)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/util/compat.py", line 79, in load_module_py
    mod = imp.load_source(module_id, path, fp)
  File "CONST_alembic/main/env.py", line 110, in <module>
    run_migrations_online()
  File "CONST_alembic/main/env.py", line 102, in run_migrations_online
    context.run_migrations()
  File "<string>", line 8, in run_migrations
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/runtime/environment.py", line 797, in run_migrations
    self.get_context().run_migrations(**kw)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/runtime/migration.py", line 312, in run_migrations
    step.migration_fn(**kw)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/CONST_alembic/main/versions/e004f76e951a_add_missing_not_null.py", line 51, in upgrade
    op.alter_column('layer_wmts', 'image_type', nullable=False, schema=schema)
  File "<string>", line 8, in alter_column
  File "<string>", line 3, in alter_column
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/operations/ops.py", line 1420, in alter_column
    return operations.invoke(alt)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/operations/base.py", line 318, in invoke
    return fn(self, operation)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/operations/toimpl.py", line 53, in alter_column
    **operation.kw
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/ddl/postgresql.py", line 105, in alter_column
    **kw)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/ddl/impl.py", line 145, in alter_column
    existing_nullable=existing_nullable,
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/alembic/ddl/impl.py", line 118, in _exec
    return conn.execute(construct, *multiparams, **params)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 945, in execute
    return meth(self, multiparams, params)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/sqlalchemy/sql/ddl.py", line 68, in _execute_on_connection
    return connection._execute_ddl(self, multiparams, params)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1002, in _execute_ddl
    compiled
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1189, in _execute_context
    context)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1393, in _handle_dbapi_exception
    exc_info
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 202, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
    context)
  File "/var/www/vhosts/cartoriviera_preprod/private/cartoriviera_geomapfish_v2/cartoriviera_geomapfish/.build/venv/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 462, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (psycopg2.IntegrityError) column "image_type" contains null values
 [SQL: 'ALTER TABLE main.layer_wmts ALTER COLUMN image_type SET NOT NULL']
CONST_Makefile:1256: recipe for target 'upgrade4' failed

Now the step 4 of upgrade is done. But I have an error in step 5... : 

 make -f v2.mk upgrade5
CONST_Makefile:1148: warning: overriding recipe for target 'apache/mapcache.xml'
cartoriviera.mk:18: warning: ignoring old recipe for target 'apache/mapcache.xml'
.build/venv/bin/c2ctool upgrade  v2.mk  --step 5
================================================================
Checker error:
Open `http://preprod.cartoriviera.ch/v2/wsgi/check_collector?` for more informations.

Correct the checker, the step 5 again:
make -f v2.mk upgrade5
CONST_Makefile:1260: recipe for target 'upgrade5' failed
make: *** [upgrade5] Error 1

--- a/c2cgeoportal/scripts/themev1tov2.py
+++ b/c2cgeoportal/scripts/themev1tov2.py
@@ -192,13 +192,6 @@ def layer_v1tov2(session, layer):
     new_layer.geo_table = layer.geo_table
     new_layer.interfaces = layer.interfaces
 
-    for link in layer.parents_relation:
-        new_link = LayergroupTreeitem()
-        new_link.ordering = link.ordering
-        new_link.treegroup_id = link.treegroup_id
-        new_link.group = link.group
-        new_link.item = new_layer
-
     if layer.layer_type[-4:] == " WMS":
         new_layer.layer = layer.name
         new_layer.style = layer.style
@@ -216,6 +209,13 @@ def layer_v1tov2(session, layer):
             dimensions = loads(layer.dimensions)
             for name, value in dimensions.items():
                 session.add(Dimension(name, value, new_layer))
+
+    for link in layer.parents_relation:
+        new_link = LayergroupTreeitem()
+        new_link.ordering = link.ordering
+        new_link.treegroup_id = link.treegroup_id
+        new_link.group = link.group
+        new_link.item = new_layer
 
     layer_add_metadata(layer, new_layer, session)
 

